lib/metasploit/framework/compiler/mingw.rb

```diff
@@ -77,18 +117,21 @@ def cleanup_files
         class X86
           include Mingw
 
-          attr_reader :file_name, :keep_exe, :keep_src, :strip_syms, :link_script, :opt_lvl, :mingw_bin, :compile_options, :show_compile_cmd
+          attr_reader :file_name, :keep_exe, :keep_src, :strip_syms, :link_script, :opt_lvl, :compile_options, :show_compile_cmd, :outfile, :link_options
+          attr_accessor :mingw_bin
 
           def initialize(opts={})
             @file_name = opts[:f_name]
+            @outfile = opts[:outfile]
             @keep_exe = opts[:keep_exe]
             @keep_src = opts[:keep_src]
             @strip_syms = opts[:strip_symbols]
             @show_compile_cmd = opts[:show_compile_cmd]
             @link_script = opts[:linker_script]
             @compile_options = opts[:compile_options]
+            @link_options = opts[:link_options]
             @opt_lvl = opts[:opt_lvl]
-            @mingw_bin = MINGW_X86
+            @mingw_bin = opts[:mingw_bin] || MINGW_X86
```

Comment:
Not sure why you are changing this given this is meant to be for generating a x86 binary? Mind explaining what your intent is here? Same comment applies for changing `@mingw_bin` below on the `X64` class.

Reply:
Those were set as the default hard-coded compilers when the module only compiled C. Now that it can handle both cpp and c, I made it assignable, but defaulted to the previous defaults so as not to impact existing modules that use this compiler module. Really the only options should be 'i686-w64-mingw32-gcc' or `'i686-w64-mingw32-g++'`.

A lot of this code i'm not particularly proud of since a lot of my decisions were based around not breaking existing modules.

Reply:
Ah okay makes sense. I was under the impression that MINGW_X86 would also be applicable to the CPP compilation, but it appears that might not be the case. That being said if it is only two set options it might be better to just assign some sort of flag to switch between the two vs giving arbitrary control to the caller and placing the burden on them to know what to supply.

Just a thought but let me know what you think 👍

Reply:
I like that. Do you think selecting the compiler by file extension is too fragile? This would remove the option of selecting completely, and I could also choose the ARCH for them as well based on the PAYLOAD selected.

```ruby
# psuedocode
compiler_map = { 
    cpp: {
        x86: MINGW_CPP_X86 , 
        x64: MINGW_CPP_X64 
    }, 
    c: {
        x86: MINGW_X86, 
        x64: MINGW_X64) 
}

compiler = compiler_map[outfile_extention][payload.arch.first]
```